### PR TITLE
feat(storage): add support for window aggregate queries

### PR DESCRIPTION
**Pull request by danxmoran**
Pre-work needed to backport Flux pushdowns

The LOCs added here is huge, but it should almost all be a 1:1 match to the code that's in 2.x. I'll try to find some `git diff` commands to show that things match, to help with review.

### CODE REVIEW

File: storage/reads/aggregate_resultset.go [hunk 1]
```diff
@@ -0,0 +1,141 @@
+package reads
```
Comment (danxmoran) [inline storage/reads/aggregate_resultset.go:1]:
```
$ git diff master -- storage/reads/aggregate_resultset.go
diff --git a/storage/reads/aggregate_resultset.go b/storage/reads/aggregate_resultset.go
index dc2694bd2d..82a15d1e20 100644
--- a/storage/reads/aggregate_resultset.go
+++ b/storage/reads/aggregate_resultset.go
@@ -6,11 +6,10 @@ import (

        "github.com/influxdata/flux/interval"
        "github.com/influxdata/flux/values"
-       "github.com/influxdata/influxdb/v2/kit/errors"
-       "github.com/influxdata/influxdb/v2/kit/tracing"
-       "github.com/influxdata/influxdb/v2/models"
-       "github.com/influxdata/influxdb/v2/storage/reads/datatypes"
-       "github.com/influxdata/influxdb/v2/tsdb/cursors"
+       "github.com/influxdata/influxdb/kit/errors"
+       "github.com/influxdata/influxdb/models"
+       "github.com/influxdata/influxdb/storage/reads/datatypes"
+       "github.com/influxdata/influxdb/tsdb/cursors"
 )

 type windowAggregateResultSet struct {
@@ -24,14 +23,6 @@ type windowAggregateResultSet struct {
 }

 func NewWindowAggregateResultSet(ctx context.Context, req *datatypes.ReadWindowAggregateRequest, cursor SeriesCursor) (ResultSet, error) {
-       span, _ := tracing.StartSpanFromContext(ctx)
-       defer span.Finish()
-
-       span.LogKV("aggregate_window_every", req.WindowEvery)
-       for _, aggregate := range req.Aggregate {
-               span.LogKV("aggregate_type", aggregate.String())
-       }
-
        if nAggs := len(req.Aggregate); nAggs != 1 {
                return nil, errors.Errorf(errors.InternalError, "attempt to create a windowAggregateResultSet with %v aggregate functions", nAggs)
        }
```

File: storage/reads/array_cursor.gen.go [hunk 2]
```diff
@@ -8,7 +8,11 @@ package reads
```
Comment (danxmoran) [inline storage/reads/array_cursor.gen.go:8]: This file is generated from the matching `.tmpl`, see there for any comments.

File: storage/reads/array_cursor.gen.go.tmpl [hunk 3]
```diff
@@ -0,0 +1,581 @@
+package reads
```
Comment (danxmoran) [inline storage/reads/array_cursor.gen.go.tmpl:1]:
```
$ git diff master -- storage/reads/array_cursor.gen.go.tmpl
diff --git a/storage/reads/array_cursor.gen.go.tmpl b/storage/reads/array_cursor.gen.go.tmpl
index a1c3f27888..428310dec7 100644
--- a/storage/reads/array_cursor.gen.go.tmpl
+++ b/storage/reads/array_cursor.gen.go.tmpl
@@ -7,8 +7,8 @@ import (

     "github.com/influxdata/flux/interval"
     "github.com/influxdata/flux/values"
-    errors2 "github.com/influxdata/influxdb/v2/kit/platform/errors"
-       "github.com/influxdata/influxdb/v2/tsdb/cursors"
+       errors2 "github.com/influxdata/influxdb/kit/platform/errors"
+       "github.com/influxdata/influxdb/tsdb/cursors"
 )

 const (
```

File: storage/reads/array_cursor.gen.go.tmpldata [hunk 4]
```diff
@@ -0,0 +1,173 @@
+[
```
Comment (danxmoran) [inline storage/reads/array_cursor.gen.go.tmpldata:1]: Identical to 2.x

File: storage/reads/array_cursor.go [hunk 5]
```diff
@@ -4,6 +4,7 @@ import (
 	"context"
```
Comment (danxmoran) [inline storage/reads/array_cursor.go:4]:
```
$ git diff master -- storage/reads/array_cursor.go
diff --git a/storage/reads/array_cursor.go b/storage/reads/array_cursor.go
index 3e9f580b11..5ab63a5972 100644
--- a/storage/reads/array_cursor.go
+++ b/storage/reads/array_cursor.go
@@ -5,8 +5,8 @@ import (
        "fmt"

        "github.com/influxdata/flux/interval"
-       "github.com/influxdata/influxdb/v2/storage/reads/datatypes"
-       "github.com/influxdata/influxdb/v2/tsdb/cursors"
+       "github.com/influxdata/influxdb/storage/reads/datatypes"
+       "github.com/influxdata/influxdb/tsdb/cursors"
 )

 type singleValue struct {
```

File: storage/reads/array_cursor_gen_test.go [hunk 6]
```diff
@@ -0,0 +1,1297 @@
+// Generated by tmpl
```
Comment (danxmoran) [inline storage/reads/array_cursor_gen_test.go:1]: Generated from the tmpl, see there for comments

File: storage/reads/array_cursor_test.gen.go.tmpl [hunk 7]
```diff
@@ -0,0 +1,115 @@
+package reads
```
Comment (danxmoran) [inline storage/reads/array_cursor_test.gen.go.tmpl:1]:
```
$ git diff master -- storage/reads/array_cursor_test.gen.go.tmpl
diff --git a/storage/reads/array_cursor_test.gen.go.tmpl b/storage/reads/array_cursor_test.gen.go.tmpl
index 2500352346..a453e83823 100644
--- a/storage/reads/array_cursor_test.gen.go.tmpl
+++ b/storage/reads/array_cursor_test.gen.go.tmpl
@@ -8,8 +8,8 @@ import (
        "github.com/influxdata/flux/interval"
        "github.com/influxdata/flux/values"
        "github.com/google/go-cmp/cmp"
-       "github.com/influxdata/influxdb/v2/tsdb/cursors"
-       "github.com/influxdata/influxdb/v2/storage/reads/datatypes"
+       "github.com/influxdata/influxdb/tsdb/cursors"
+       "github.com/influxdata/influxdb/storage/reads/datatypes"
 )

 var cmpOptions = cmp.AllowUnexported(interval.Window{})
```

File: storage/reads/datatypes/storage_common.pb.go [hunk 8]
```diff
@@ -99,18 +99,33 @@ const (
 	AggregateTypeNone  Aggregate_AggregateType = 0
```
Comment (danxmoran) [inline storage/reads/datatypes/storage_common.pb.go:99]: Generated, see the `.proto` for comments

File: storage/reads/flux_reader.go [hunk 9]
```diff
@@ -8,14 +8,43 @@ import (
 	"github.com/gogo/protobuf/types"
```
Comment (danxmoran) [inline storage/reads/flux_reader.go:8]:
```
$ git diff master:./storage/flux/reader.go ./storage/reads/flux_reader.go
diff --git a/storage/flux/reader.go b/storage/reads/flux_reader.go
index e02cabc304..b69eb290ac 100644
--- a/storage/flux/reader.go
+++ b/storage/reads/flux_reader.go
@@ -1,12 +1,10 @@
-package storageflux
+package reads
 
 import (
 	"context"
 	"fmt"
 	"strings"
 
-	"github.com/influxdata/influxdb/v2/kit/platform/errors"
-
 	"github.com/gogo/protobuf/types"
 	"github.com/influxdata/flux"
 	"github.com/influxdata/flux/execute"
@@ -14,11 +12,11 @@ import (
 	"github.com/influxdata/flux/memory"
 	"github.com/influxdata/flux/plan"
 	"github.com/influxdata/flux/values"
-	"github.com/influxdata/influxdb/v2/models"
-	"github.com/influxdata/influxdb/v2/query"
-	storage "github.com/influxdata/influxdb/v2/storage/reads"
-	"github.com/influxdata/influxdb/v2/storage/reads/datatypes"
-	"github.com/influxdata/influxdb/v2/tsdb/cursors"
+	"github.com/influxdata/influxdb/flux/stdlib/influxdata/influxdb"
+	"github.com/influxdata/influxdb/kit/platform/errors"
+	"github.com/influxdata/influxdb/models"
+	"github.com/influxdata/influxdb/storage/reads/datatypes"
+	"github.com/influxdata/influxdb/tsdb/cursors"
 )
 
 // GroupCursorError is returned when two different cursor types
@@ -55,15 +53,15 @@ type storageTable interface {
 }
 
 type storeReader struct {
-	s storage.Store
+	s Store
 }
 
 // NewReader returns a new storageflux reader
-func NewReader(s storage.Store) query.StorageReader {
+func NewReader(s Store) influxdb.Reader {
 	return &storeReader{s: s}
 }
 
-func (r *storeReader) ReadFilter(ctx context.Context, spec query.ReadFilterSpec, alloc *memory.Allocator) (query.TableIterator, error) {
+func (r *storeReader) ReadFilter(ctx context.Context, spec influxdb.ReadFilterSpec, alloc *memory.Allocator) (influxdb.TableIterator, error) {
 	return &filterIterator{
 		ctx:   ctx,
 		s:     r.s,
@@ -73,7 +71,7 @@ func (r *storeReader) ReadFilter(ctx context.Context, spec query.ReadFilterSpec,
 	}, nil
 }
 
-func (r *storeReader) ReadGroup(ctx context.Context, spec query.ReadGroupSpec, alloc *memory.Allocator) (query.TableIterator, error) {
+func (r *storeReader) ReadGroup(ctx context.Context, spec influxdb.ReadGroupSpec, alloc *memory.Allocator) (influxdb.TableIterator, error) {
 	return &groupIterator{
 		ctx:   ctx,
 		s:     r.s,
@@ -83,7 +81,7 @@ func (r *storeReader) ReadGroup(ctx context.Context, spec query.ReadGroupSpec, a
 	}, nil
 }
 
-func (r *storeReader) ReadWindowAggregate(ctx context.Context, spec query.ReadWindowAggregateSpec, alloc *memory.Allocator) (query.TableIterator, error) {
+func (r *storeReader) ReadWindowAggregate(ctx context.Context, spec influxdb.ReadWindowAggregateSpec, alloc *memory.Allocator) (influxdb.TableIterator, error) {
 	return &windowAggregateIterator{
 		ctx:   ctx,
 		s:     r.s,
@@ -93,7 +91,7 @@ func (r *storeReader) ReadWindowAggregate(ctx context.Context, spec query.ReadWi
 	}, nil
 }
 
-func (r *storeReader) ReadTagKeys(ctx context.Context, spec query.ReadTagKeysSpec, alloc *memory.Allocator) (query.TableIterator, error) {
+func (r *storeReader) ReadTagKeys(ctx context.Context, spec influxdb.ReadTagKeysSpec, alloc *memory.Allocator) (influxdb.TableIterator, error) {
 	return &tagKeysIterator{
 		ctx:       ctx,
 		bounds:    spec.Bounds,
@@ -104,7 +102,7 @@ func (r *storeReader) ReadTagKeys(ctx context.Context, spec query.ReadTagKeysSpe
 	}, nil
 }
 
-func (r *storeReader) ReadTagValues(ctx context.Context, spec query.ReadTagValuesSpec, alloc *memory.Allocator) (query.TableIterator, error) {
+func (r *storeReader) ReadTagValues(ctx context.Context, spec influxdb.ReadTagValuesSpec, alloc *memory.Allocator) (influxdb.TableIterator, error) {
 	return &tagValuesIterator{
 		ctx:       ctx,
 		bounds:    spec.Bounds,
@@ -119,8 +117,8 @@ func (r *storeReader) Close() {}
 
 type filterIterator struct {
 	ctx   context.Context
-	s     storage.Store
-	spec  query.ReadFilterSpec
+	s     Store
+	spec  influxdb.ReadFilterSpec
 	stats cursors.CursorStats
 	cache *tagsCache
 	alloc *memory.Allocator
@@ -130,8 +128,8 @@ func (fi *filterIterator) Statistics() cursors.CursorStats { return fi.stats }
 
 func (fi *filterIterator) Do(f func(flux.Table) error) error {
 	src := fi.s.GetSource(
-		uint64(fi.spec.OrganizationID),
-		uint64(fi.spec.BucketID),
+		fi.spec.Database,
+		fi.spec.RetentionPolicy,
 	)
 
 	// Setup read request
@@ -158,7 +156,7 @@ func (fi *filterIterator) Do(f func(flux.Table) error) error {
 	return fi.handleRead(f, rs)
 }
 
-func (fi *filterIterator) handleRead(f func(flux.Table) error, rs storage.ResultSet) error {
+func (fi *filterIterator) handleRead(f func(flux.Table) error, rs ResultSet) error {
 	// these resources must be closed if not nil on return
 	var (
 		cur   cursors.Cursor
@@ -234,8 +232,8 @@ READ:
 
 type groupIterator struct {
 	ctx   context.Context
-	s     storage.Store
-	spec  query.ReadGroupSpec
+	s     Store
+	spec  influxdb.ReadGroupSpec
 	stats cursors.CursorStats
 	cache *tagsCache
 	alloc *memory.Allocator
@@ -245,8 +243,8 @@ func (gi *groupIterator) Statistics() cursors.CursorStats { return gi.stats }
 
 func (gi *groupIterator) Do(f func(flux.Table) error) error {
 	src := gi.s.GetSource(
-		uint64(gi.spec.OrganizationID),
-		uint64(gi.spec.BucketID),
+		gi.spec.Database,
+		gi.spec.RetentionPolicy,
 	)
 
 	// Setup read request
@@ -261,7 +259,7 @@ func (gi *groupIterator) Do(f func(flux.Table) error) error {
 	req.Range.Start = int64(gi.spec.Bounds.Start)
 	req.Range.End = int64(gi.spec.Bounds.Stop)
 
-	if len(gi.spec.GroupKeys) > 0 && gi.spec.GroupMode == query.GroupModeNone {
+	if len(gi.spec.GroupKeys) > 0 && gi.spec.GroupMode == influxdb.GroupModeNone {
 		return &errors.Error{
 			Code: errors.EInternal,
 			Msg:  "cannot have group mode none with group key values",
@@ -287,10 +285,10 @@ func (gi *groupIterator) Do(f func(flux.Table) error) error {
 	return gi.handleRead(f, rs)
 }
 
-func (gi *groupIterator) handleRead(f func(flux.Table) error, rs storage.GroupResultSet) error {
+func (gi *groupIterator) handleRead(f func(flux.Table) error, rs GroupResultSet) error {
 	// these resources must be closed if not nil on return
 	var (
-		gc    storage.GroupCursor
+		gc    GroupCursor
 		cur   cursors.Cursor
 		table storageTable
 	)
@@ -386,11 +384,11 @@ func determineAggregateMethod(agg string) (datatypes.Aggregate_AggregateType, er
 	return 0, fmt.Errorf("unknown aggregate type %q", agg)
 }
 
-func convertGroupMode(m query.GroupMode) datatypes.ReadGroupRequest_Group {
+func convertGroupMode(m influxdb.GroupMode) datatypes.ReadGroupRequest_Group {
 	switch m {
-	case query.GroupModeNone:
+	case influxdb.GroupModeNone:
 		return datatypes.GroupNone
-	case query.GroupModeBy:
+	case influxdb.GroupModeBy:
 		return datatypes.GroupBy
 	}
 	panic(fmt.Sprint("invalid group mode: ", m))
@@ -577,7 +575,7 @@ func determineTableColsForGroup(tagKeys [][]byte, typ flux.ColType, agg *datatyp
 	return cols, defs
 }
 
-func groupKeyForGroup(kv [][]byte, spec *query.ReadGroupSpec, bnds execute.Bounds) flux.GroupKey {
+func groupKeyForGroup(kv [][]byte, spec *influxdb.ReadGroupSpec, bnds execute.Bounds) flux.GroupKey {
 	cols := make([]flux.ColMeta, 2, len(spec.GroupKeys)+2)
 	vs := make([]values.Value, 2, len(spec.GroupKeys)+2)
 	cols[startColIdx] = flux.ColMeta{
@@ -605,8 +603,8 @@ func groupKeyForGroup(kv [][]byte, spec *query.ReadGroupSpec, bnds execute.Bound
 
 type windowAggregateIterator struct {
 	ctx   context.Context
-	s     storage.Store
-	spec  query.ReadWindowAggregateSpec
+	s     Store
+	spec  influxdb.ReadWindowAggregateSpec
 	stats cursors.CursorStats
 	cache *tagsCache
 	alloc *memory.Allocator
@@ -616,8 +614,8 @@ func (wai *windowAggregateIterator) Statistics() cursors.CursorStats { return wa
 
 func (wai *windowAggregateIterator) Do(f func(flux.Table) error) error {
 	src := wai.s.GetSource(
-		uint64(wai.spec.OrganizationID),
-		uint64(wai.spec.BucketID),
+		wai.spec.Database,
+		wai.spec.RetentionPolicy,
 	)
 
 	// Setup read request
@@ -681,7 +679,7 @@ func isSelector(kind plan.ProcedureKind) bool {
 	return kind == FirstKind || kind == LastKind || kind == MinKind || kind == MaxKind
 }
 
-func (wai *windowAggregateIterator) handleRead(f func(flux.Table) error, rs storage.ResultSet) error {
+func (wai *windowAggregateIterator) handleRead(f func(flux.Table) error, rs ResultSet) error {
 	createEmpty := wai.spec.CreateEmpty
 
 	selector := len(wai.spec.Aggregates) > 0 && isSelector(wai.spec.Aggregates[0])
@@ -839,16 +837,16 @@ func isAggregateCount(kind plan.ProcedureKind) bool {
 type tagKeysIterator struct {
 	ctx       context.Context
 	bounds    execute.Bounds
-	s         storage.Store
-	readSpec  query.ReadTagKeysSpec
+	s         Store
+	readSpec  influxdb.ReadTagKeysSpec
 	predicate *datatypes.Predicate
 	alloc     *memory.Allocator
 }
 
 func (ti *tagKeysIterator) Do(f func(flux.Table) error) error {
 	src := ti.s.GetSource(
-		uint64(ti.readSpec.OrganizationID),
-		uint64(ti.readSpec.BucketID),
+		ti.readSpec.Database,
+		ti.readSpec.RetentionPolicy,
 	)
 
 	var req datatypes.TagKeysRequest
@@ -922,16 +920,16 @@ func (ti *tagKeysIterator) Statistics() cursors.CursorStats {
 type tagValuesIterator struct {
 	ctx       context.Context
 	bounds    execute.Bounds
-	s         storage.Store
-	readSpec  query.ReadTagValuesSpec
+	s         Store
+	readSpec  influxdb.ReadTagValuesSpec
 	predicate *datatypes.Predicate
 	alloc     *memory.Allocator
 }
 
 func (ti *tagValuesIterator) Do(f func(flux.Table) error) error {
 	src := ti.s.GetSource(
-		uint64(ti.readSpec.OrganizationID),
-		uint64(ti.readSpec.BucketID),
+		ti.readSpec.Database,
+		ti.readSpec.RetentionPolicy,
 	)
 
 	var req datatypes.TagValuesRequest
```

File: storage/reads/flux_table.gen.go [hunk 10]
```diff
@@ -7,15 +7,21 @@
 package reads
```
Comment (danxmoran) [inline storage/reads/flux_table.gen.go:7]: Generated, see the `.tmpl` for comments

File: storage/reads/flux_table.gen.go.tmpl [hunk 11]
```diff
@@ -0,0 +1,945 @@
+package reads
```
Comment (danxmoran) [inline storage/reads/flux_table.gen.go.tmpl:1]:
```
$ git diff master:./storage/flux/table.gen.go.tmpl ./storage/reads/flux_table.gen.go.tmpl
diff --git a/storage/flux/table.gen.go.tmpl b/storage/reads/flux_table.gen.go.tmpl
index e6c168967c..84e0a5094c 100644
--- a/storage/flux/table.gen.go.tmpl
+++ b/storage/reads/flux_table.gen.go.tmpl
@@ -1,4 +1,4 @@
-package storageflux 
+package reads
 
 import (
 	"fmt"
@@ -12,11 +12,10 @@ import (
 	"github.com/influxdata/flux/interval"
 	"github.com/influxdata/flux/memory"
 	"github.com/influxdata/flux/values"
-	"github.com/influxdata/influxdb/v2/kit/platform/errors"
-	"github.com/influxdata/influxdb/v2/models"
-	"github.com/influxdata/influxdb/v2/storage/reads/datatypes"
-	storage "github.com/influxdata/influxdb/v2/storage/reads"
-	"github.com/influxdata/influxdb/v2/tsdb/cursors"
+	"github.com/influxdata/influxdb/kit/platform/errors"
+	"github.com/influxdata/influxdb/models"
+	"github.com/influxdata/influxdb/storage/reads/datatypes"
+	"github.com/influxdata/influxdb/tsdb/cursors"
 )
 {{range .}}
 //
@@ -163,8 +162,8 @@ func (t *{{.name}}WindowTable) createNextBufferTimes() (start, stop *array.Int64
 
 		// Create a buffer with the buffer size.
 		// TODO(jsternberg): Calculate the exact size with max points as the maximum.
-		startB.Resize(storage.MaxPointsPerBlock)
-		stopB.Resize(storage.MaxPointsPerBlock)
+		startB.Resize(MaxPointsPerBlock)
+		stopB.Resize(MaxPointsPerBlock)
 		for ; ; t.windowBounds = t.window.NextBounds(t.windowBounds) {
 			startT, stopT := t.getWindowBoundsFor(t.windowBounds)
 			if startT >= int64(t.bounds.Stop) {
@@ -323,7 +322,7 @@ func new{{.Name}}WindowSelectorTable(
 	done chan struct{},
 	cur cursors.{{.Name}}ArrayCursor,
 	bounds execute.Bounds,
-	window interval.Window, 
+	window interval.Window,
 	timeColumn string,
 	key flux.GroupKey,
 	cols []flux.ColMeta,
@@ -443,7 +442,7 @@ func new{{.Name}}EmptyWindowSelectorTable(
 		arr: cur.Next(),
 		rangeStart:  rangeStart,
 		rangeStop:   rangeStop,
-		windowBounds: window.GetLatestBounds(values.Time(rangeStart)),
+        windowBounds: window.GetLatestBounds(values.Time(rangeStart)),
 		window: window,
 		timeColumn:  timeColumn,
 	}
@@ -462,7 +461,7 @@ func (t *{{.name}}EmptyWindowSelectorTable) advance() bool {
 	}
 
 	values := t.arrowBuilder()
-	values.Resize(storage.MaxPointsPerBlock)
+	values.Resize(MaxPointsPerBlock)
 
 	var cr *colReader
 
@@ -492,7 +491,7 @@ func (t *{{.name}}EmptyWindowSelectorTable) advance() bool {
 
 func (t *{{.name}}EmptyWindowSelectorTable) startTimes(builder *array.{{.ArrowType}}Builder) *array.Int64 {
 	start := arrow.NewIntBuilder(t.alloc)
-	start.Resize(storage.MaxPointsPerBlock)
+	start.Resize(MaxPointsPerBlock)
 
 	for int64(t.windowBounds.Start()) < t.rangeStop {
 		// The first window should start at the
@@ -530,7 +529,7 @@ func (t *{{.name}}EmptyWindowSelectorTable) startTimes(builder *array.{{.ArrowTy
 			t.idx = 0
 		}
 
-		if start.Len() == storage.MaxPointsPerBlock {
+		if start.Len() == MaxPointsPerBlock {
 			break
 		}
 	}
@@ -539,7 +538,7 @@ func (t *{{.name}}EmptyWindowSelectorTable) startTimes(builder *array.{{.ArrowTy
 
 func (t *{{.name}}EmptyWindowSelectorTable) stopTimes(builder *array.{{.ArrowType}}Builder) *array.Int64 {
 	stop := arrow.NewIntBuilder(t.alloc)
-	stop.Resize(storage.MaxPointsPerBlock)
+	stop.Resize(MaxPointsPerBlock)
 
 	for int64(t.windowBounds.Start()) < t.rangeStop {
 		// The last window should stop at the end of
@@ -577,7 +576,7 @@ func (t *{{.name}}EmptyWindowSelectorTable) stopTimes(builder *array.{{.ArrowTyp
 			t.idx = 0
 		}
 
-		if stop.Len() == storage.MaxPointsPerBlock {
+		if stop.Len() == MaxPointsPerBlock {
 			break
 		}
 	}
@@ -586,13 +585,13 @@ func (t *{{.name}}EmptyWindowSelectorTable) stopTimes(builder *array.{{.ArrowTyp
 
 func (t *{{.name}}EmptyWindowSelectorTable) startStopTimes(builder *array.{{.ArrowType}}Builder) (*array.Int64, *array.Int64, *array.Int64) {
 	start := arrow.NewIntBuilder(t.alloc)
-	start.Resize(storage.MaxPointsPerBlock)
+	start.Resize(MaxPointsPerBlock)
 
 	stop := arrow.NewIntBuilder(t.alloc)
-	stop.Resize(storage.MaxPointsPerBlock)
+	stop.Resize(MaxPointsPerBlock)
 
 	time := arrow.NewIntBuilder(t.alloc)
-	time.Resize(storage.MaxPointsPerBlock)
+	time.Resize(MaxPointsPerBlock)
 
 	for int64(t.windowBounds.Start()) < t.rangeStop {
 
@@ -641,7 +640,7 @@ func (t *{{.name}}EmptyWindowSelectorTable) startStopTimes(builder *array.{{.Arr
 			t.idx = 0
 		}
 
-		if time.Len() == storage.MaxPointsPerBlock {
+		if time.Len() == MaxPointsPerBlock {
 			break
 		}
 	}
@@ -653,13 +652,13 @@ func (t *{{.name}}EmptyWindowSelectorTable) startStopTimes(builder *array.{{.Arr
 type {{.name}}GroupTable struct {
 	table
 	mu     sync.Mutex
-	gc     storage.GroupCursor
+	gc     GroupCursor
 	cur    cursors.{{.Name}}ArrayCursor
 }
 
 func new{{.Name}}GroupTable(
 	done chan struct{},
-	gc storage.GroupCursor,
+	gc GroupCursor,
 	cur cursors.{{.Name}}ArrayCursor,
 	bounds execute.Bounds,
 	key flux.GroupKey,
@@ -920,7 +919,7 @@ func (t *{{.name}}GroupTable) advanceCursor() bool {
 				Err: &GroupCursorError {
 					typ: "{{.name}}",
 					cursor: cur,
-				},
+                },
 			}
 			return false
 		} else {

```

File: storage/reads/flux_table.go [hunk 12]
```diff
@@ -1,5 +1,7 @@
 package reads
```
Comment (danxmoran) [inline storage/reads/flux_table.go:1]:
```
$ git diff master:./storage/flux/table.go ./storage/reads/flux_table.go
diff --git a/storage/flux/table.go b/storage/reads/flux_table.go
index 5173802447..a13ebfbb8b 100644
--- a/storage/flux/table.go
+++ b/storage/reads/flux_table.go
@@ -1,6 +1,6 @@
-package storageflux
+package reads
 
-//go:generate env GO111MODULE=on go run github.com/benbjohnson/tmpl -data=@types.tmpldata table.gen.go.tmpl
+//go:generate env GO111MODULE=on go run github.com/benbjohnson/tmpl -data=@types.tmpldata flux_table.gen.go.tmpl
 
 import (
 	"errors"
@@ -11,7 +11,7 @@ import (
 	"github.com/influxdata/flux/arrow"
 	"github.com/influxdata/flux/execute"
 	"github.com/influxdata/flux/memory"
-	"github.com/influxdata/influxdb/v2/models"
+	"github.com/influxdata/influxdb/models"
 )
 
 type table struct {
```

File: storage/reads/group_resultset.go [hunk 13]
```diff
@@ -4,7 +4,6 @@ import (
 	"bytes"
```
Comment (danxmoran) [inline storage/reads/group_resultset.go:4]:
```
$ git diff master ./storage/reads/group_resultset.go
diff --git a/storage/reads/group_resultset.go b/storage/reads/group_resultset.go
index d3bf2ca12d..c0d4c04e4a 100644
--- a/storage/reads/group_resultset.go
+++ b/storage/reads/group_resultset.go
@@ -6,9 +6,9 @@ import (
        "fmt"
        "sort"

-       "github.com/influxdata/influxdb/v2/models"
-       "github.com/influxdata/influxdb/v2/storage/reads/datatypes"
-       "github.com/influxdata/influxdb/v2/tsdb/cursors"
+       "github.com/influxdata/influxdb/models"
+       "github.com/influxdata/influxdb/storage/reads/datatypes"
+       "github.com/influxdata/influxdb/tsdb/cursors"
 )

 type groupResultSet struct {
```

File: storage/reads/resultset.go [hunk 14]
```diff
@@ -2,31 +2,27 @@ package reads
```
Comment (danxmoran) [inline storage/reads/resultset.go:2]:
```
$ git diff master ./storage/reads/resultset.go
diff --git a/storage/reads/resultset.go b/storage/reads/resultset.go
index 9f7d575504..dc6ba9405d 100644
--- a/storage/reads/resultset.go
+++ b/storage/reads/resultset.go
@@ -3,8 +3,8 @@ package reads
 import (
        "context"

-       "github.com/influxdata/influxdb/v2/models"
-       "github.com/influxdata/influxdb/v2/tsdb/cursors"
+       "github.com/influxdata/influxdb/models"
+       "github.com/influxdata/influxdb/tsdb/cursors"
 )

 type multiShardCursors interface {
```

File: storage/reads/window.go [hunk 15]
```diff
@@ -0,0 +1,199 @@
+package reads
```
Comment (danxmoran) [inline storage/reads/window.go:1]:
```
$ git diff master:./storage/flux/window.go ./storage/reads/window.go
diff --git a/storage/flux/window.go b/storage/reads/window.go
index b9b3675bee..b9db96b103 100644
--- a/storage/flux/window.go
+++ b/storage/reads/window.go
@@ -1,18 +1,17 @@
-package storageflux
+package reads

 import (
        "context"
        "fmt"
        "sync/atomic"

-       "github.com/influxdata/influxdb/v2/kit/platform/errors"
-
        "github.com/apache/arrow/go/arrow/array"
        "github.com/apache/arrow/go/arrow/memory"
        "github.com/influxdata/flux"
        "github.com/influxdata/flux/arrow"
        "github.com/influxdata/flux/execute"
        "github.com/influxdata/flux/values"
+       "github.com/influxdata/influxdb/kit/platform/errors"
 )

 // splitWindows will split a windowTable by creating a new table from each
```

File: go.mod [hunk 16]
```diff
@@ -6,6 +6,7 @@ require (
 	collectd.org v0.3.0
 	github.com/BurntSushi/toml v0.3.1
 	github.com/apache/arrow/go/arrow v0.0.0-20200923215132-ac86123a3f01
+	github.com/benbjohnson/tmpl v1.0.0
```
Comment (lesam) [inline go.mod:9]: 👍

File: services/storage/store.go [hunk 17]
```diff
@@ -188,6 +188,41 @@ func (s *Store) ReadGroup(ctx context.Context, req *datatypes.ReadGroupRequest)
...
+	database, rp, start, end, err := s.validateArgs(source.Database, source.RetentionPolicy, req.Range.Start, req.Range.End)
+	if err != nil {
+		return nil, err
+	}
+
+	shardIDs, err := s.findShardIDs(database, rp, false, start, end)
+	if err != nil {
+		return nil, err
+	}
+	if len(shardIDs) == 0 { // TODO(jeff): this was a typed nil
```
[the remark this answers is not in the document]
Comment (lesam) [reply services/storage/store.go:210]: I see we explicitly check for nil in windowAggregateIterator.Do - interesting.

File: storage/reads/array_cursor.gen.go.tmpl (hunk 3 above)
Comment (lesam) [reply storage/reads/array_cursor.gen.go.tmpl:1]: It looks like this file is new - did we copy the .gen.go without copying the .gen.go.tmpl before?

File: storage/reads/array_cursor.gen.go.tmpl [hunk 18]
```diff
@@ -0,0 +1,588 @@
...
+    "github.com/influxdata/flux/values"
+	errors2 "github.com/influxdata/influxdb/kit/platform/errors"
+	"github.com/influxdata/influxdb/tsdb/cursors"
+)
+
+const (
+	// MaxPointsPerBlock is the maximum number of points in an encoded
+	// block in a TSM file. It should match the value in the tsm1
+	// package, but we don't want to import it.
+	MaxPointsPerBlock = 1000
```
Comment (lesam) [inline storage/reads/array_cursor.gen.go.tmpl:18]: Interestingly we have a MaxPointsPerBlock in the Engine struct in tsm1, but we don't use it, we only use DefaultMaxPointsPerBlock

File: storage/reads/array_cursor.gen.go.tmpl (hunk 3 above)
Comment (danxmoran) [reply storage/reads/array_cursor.gen.go.tmpl:1]: Yes (at least as far as I can tell from the git history)

File: services/storage/store.go (hunk 17 above)
[the remark this answers is not in the document]
Comment (danxmoran) [reply services/storage/store.go:210]: I'll see what the 2.x git history has to say

File: storage/reads/flux_reader.go (hunk 9 above)
Comment (lesam) [reply storage/reads/flux_reader.go:8]: Might be worth renaming the `influxdb` import to `query` for easier backporting

File: storage/reads/flux_table.gen.go.tmpl (hunk 11 above)
Comment (lesam) [reply storage/reads/flux_table.gen.go.tmpl:1]: A few whitespace diffs are annoying.

Comment (lesam) [reply storage/reads/flux_table.gen.go.tmpl:1]: Also it's worrying that this `tmpl` didn't exist but the generated code was here?

File: storage/reads/window.go [hunk 19]
```diff
@@ -0,0 +1,199 @@
...
+}
+
+type windowTableSplitter struct {
+	ctx      context.Context
+	in       flux.Table
+	alloc    memory.Allocator
+	selector bool
+}
+
+func (w *windowTableSplitter) Do(f func(flux.Table) error) error {
```
Comment (lesam) [inline storage/reads/window.go:37]: I'm a bit surprised I don't see `alloc` being used here

**Review by lesam — COMMENTED** [review]
LGTM but I think it would be valuable for someone who knows the context of the changes here (maybe @fchikwekwe ) to have a second look.

File: storage/reads/flux_reader.go (hunk 9 above)
Comment (danxmoran) [reply storage/reads/flux_reader.go:8]: I'm going to hold off for now and do all the renaming / rearranging as part of #21112

File: storage/reads/flux_tags_cache_test.go [hunk 20]
```diff
@@ -0,0 +1,130 @@
+package reads
```
Comment (danxmoran) [inline storage/reads/flux_tags_cache_test.go:1]: Identical to 2.x except for the file name and package name

File: services/storage/store.go (hunk 17 above)
[the remark this answers is not in the document]
Comment (fchikwekwe) [reply services/storage/store.go:210]: Off the top of head, I think this was about returning an actual nil ResultSet, not an empty one.  But I'm a little less familiar with this particular file.

File: storage/reads/flux_table.gen.go.tmpl (hunk 11 above)
Comment (danxmoran) [reply storage/reads/flux_table.gen.go.tmpl:1]: Whitespace differences addressed

File: storage/reads/array_cursor_test.go [hunk 21]
```diff
@@ -0,0 +1,2251 @@
+package reads
```
Comment (fchikwekwe) [inline storage/reads/array_cursor_test.go:1]: @danxmoran is this file the same as 2.x as well?

Comment (danxmoran) [reply storage/reads/array_cursor_test.go:1]: Yes (sorry, added it later). The new mock cursor files are the same too.

**Review by fchikwekwe — APPROVED** [review]
@danxmoran the additions look pretty in line with what I would expect including the import changes. There are a lot of annoying naming things that can be updated, but I saw that you're working on that in another patch. LGTM!

**Review by lesam — APPROVED** [review]

**Pull request merged**